frr-reloader/frr-reloader.sh

```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
+cleanup() {
+  echo "Caught an exit signal.."
+  rm -f "$PIDFILE"
+  rm -f "$PIPE_FILE"
+  kill_sleep
+  exit
+}
+
+reload_and_wait() {
+  reload_frr & 
+  cat "$PIPE_FILE" > /dev/null
+}
+
+reload_frr() {
+  echo "Caught SIGHUP! Reloading FRR.."
+  kill_sleep
+
+  echo "Checking the configuration file syntax"
+  if ! python3 /usr/lib/frr/frr-reload.py --test --stdout "$FILE_TO_RELOAD" ; then
+    echo "Syntax error spotted: aborting.."
+    echo "" > "$PIPE_FILE"
+    return
+  fi
+
+  echo "Applying the configuration file"
+  if ! python3 /usr/lib/frr/frr-reload.py --reload --overwrite --stdout "$FILE_TO_RELOAD" ; then
+    echo "Failed to fully apply configuration file: aborting.."
+    echo "" > "$PIPE_FILE"
+    return
+  fi
+  
+  echo "FRR reloaded successfully!"
+  echo "" > "$PIPE_FILE"
+}
+
+kill_sleep() {
+  kill "$sleep_pid"
+}
+
+trap cleanup SIGTERM SIGQUIT SIGINT
+trap reload_and_wait HUP
+
+SHARED_VOLUME="/etc/frr_reloader"
+PIDFILE="$SHARED_VOLUME/reloader.pid"
+FILE_TO_RELOAD="$SHARED_VOLUME/frr.conf"
+PIPE_FILE="$SHARED_VOLUME/done"
+
+echo "PID is: $$, writing to $PIDFILE"
+printf "$$" > "$PIDFILE"
+mkfifo "$PIPE_FILE"
+
+while true
+do
+    sleep infinity &
```

Comment:
that will behave like a ```while true``` since the wait immediately exits (it has nothing in the background to wait for). I could change it to a short sleep inside the loop instead of this wait logic if you rather, but I think resource-usage wise this is the best way (the loop isn't constantly running, wait blocks the iterations).

Reply:
I think "wait" will cause it to block and then wake up when it receives a signal. You can then exit safely from the program by calling "exit" from your trap. I tested this and it seems to work.